js/data/feature_tree.js

```diff
@@ -88,80 +152,181 @@ FeatureTree.prototype.query = function(args, styleLayersByID) {
         maxY = Math.max(maxY, p.y);
     }
 
-    var bounds = [minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius];
-    var symbolQueryBox = new CollisionBox(new Point(minX, minY), 0, 0, maxX - minX, maxY - minY, args.scale, null);
+    var matching = this.grid.query(minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius);
+    matching.sort(topDown);
+    var match = this.featureIndexArray.get(0);
+    this.filterMatching(result, matching, match, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    var matchingSymbols = this.collisionTile.queryRenderedSymbols(minX, minY, maxX, maxY, args.scale);
+    var match2 = this.collisionTile.collisionBoxArray.get(0);
+    matchingSymbols.sort();
+    this.filterMatching(result, matchingSymbols, match2, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    return result;
+};
+
+function topDown(a, b) {
+    return b - a;
+}
 
-    var matching = this.rtree.search(bounds).concat(this.collisionTile.getFeaturesAt(symbolQueryBox, args.scale));
+function getLineWidth(paint) {
+    if (paint['line-gap-width'] > 0) {
+        return paint['line-gap-width'] + 2 * paint['line-width'];
+    } else {
+        return paint['line-width'];
+    }
+}
 
+FeatureTree.prototype.filterMatching = function(result, matching, match, queryGeometry, filter, filterLayerIDs, styleLayersByID, bearing, pixelsToTileUnits, returnGeoJSON) {
+    var seen = {};
     for (var k = 0; k < matching.length; k++) {
-        var feature = matching[k].feature,
-            layerIDs = matching[k].layerIDs;
+        var index = matching[k];
 
-        if (!filter(feature)) continue;
+        if (seen[index]) continue;
+        seen[index] = true;
```

Comment:
The new mapbox streets vector tiles have huge combined multi-geometries. The entire buildings layer is one giant multipolygon. Because of this, rings are indexed individually: https://github.com/mapbox/mapbox-gl-js/blob/158227693667b2e7d6cfa2a7e5822c7d5ad81d30/js/data/feature_tree.js#L65-L79

If multiple rings intersect the query box then there will be multiple duplicate keys.

We could make Grid remove duplicates by creating an object or checking some other way, but I'm not sure that is a better spot for it.

Since the keys are in sorted order we could also just do this:
Is this better?

```diff
diff --git a/js/data/feature_tree.js b/js/data/feature_tree.js
index 954037f..c1e17ab 100644
--- a/js/data/feature_tree.js
+++ b/js/data/feature_tree.js
@@ -178,12 +178,14 @@ function getLineWidth(paint) {
 }
 
 FeatureTree.prototype.filterMatching = function(result, matching, match, queryGeometry, filter, filterLayerIDs, styleLayersByID, bearing, pixelsToTileUnits, returnGeoJSON) {
+    var previousIndex;
-    var seen = {};
     for (var k = 0; k < matching.length; k++) {
         var index = matching[k];
 
-        if (seen[index]) continue;
-        seen[index] = true;
+        // don't check the same feature more than once
+        if (index === previousIndex) continue;
+        previousIndex = index;
 
         match._setIndex(index);
```